controlplane/kubeadm/internal/etcd_client_generator.go

```diff
@@ -106,18 +111,21 @@ func (c *EtcdClientGenerator) forLeader(ctx context.Context, nodeNames []string)
 	return nil, errors.Wrap(kerrors.NewAggregate(errs), "could not establish a connection to the etcd leader")
 }
 
-func (c *EtcdClientGenerator) tryGetLeaderClient(ctx context.Context, nodeName string, allNodes sets.String) (client *etcd.Client, tryNext bool, err error) {
+// getLeaderClient provides an etcd client connected to the leader. It returns an
+// etcdNodeConnectionError if there was a connection problem with the given etcd
+// node, which should be considered non-fatal by the caller.
+func (c *EtcdClientGenerator) getLeaderClient(ctx context.Context, nodeName string, allNodes sets.String) (*etcd.Client, error) {
 	// Get a temporary client to the etcd instance hosted on the node.
-	client, err = c.forFirstAvailableNode(ctx, []string{nodeName})
+	client, err := c.forFirstAvailableNode(ctx, []string{nodeName})
 	if err != nil {
-		return nil, true, err
+		return nil, &etcdNodeConnectionError{err}
```

Comment:
```suggestion
		return nil, kerrors.NewAggregate([]errors{err, etcdNodeConnectionError})
```

I _think_ this should do it?

Reply:
This works but introduces a slight change in behavior AFAICS: any specific errors returned from `forFirstAvailableNode` are now masqueraded as `"failed to connect to etcd node"`. From an operational perspective, this may not be ideal as it may aggravate troubleshooting etcd connectivity problems when the exact error reason cannot be inspected anymore.

Personally, I'd rather see the specific error details surface, which the error type should enable. Does that make sense?

Reply:
Sorry, I overlooked the fact that your suggestion uses a new error aggregate to include both the specific error and the error variable. I'm not sure if that'd work with `errors.Is` though, and it seems odd that we now return two errors per erroring `forFirstAvailableNode` call. Or am I misunderstanding?

Reply:
Looks like `aggregate` does implement `errors.Is`. 👍 That said, I think may be a bit non-obvious to leverage `NewAggregate` so that we can use `errors.Is`. And AFAIU the logging of the error would still output two errors per failed downstream call here.

I think the error type leads to a slightly better result and could be a bit easier to understand, but let me know if you feel strong about `errors.Is` and I'm happy to adjust.

Reply:
errors.Is() should still work fine, it should be tested of course, but Aggregate just wraps all the errors

Reply:
@timoreimann I'd prefer to use the aggregate error if that's ok with you, given that we're already using it everywhere else in the codebase, it's a common expected pattern